File: examples/04-game-room/typescript/game-room.ts
/**
 * Game Room Actor Pattern - 멀티플레이어 게임 룸 예제
 *
 * 게임 룸의 Actor 기반 구현을 보여줍니다.
 */

// ============================================
// 타입 정의
// ============================================

type RoomStatus = 'waiting' | 'starting' | 'playing' | 'finished';

interface PlayerInfo {
  playerId: string;
  name: string;
  isReady: boolean;
  score: number;
}

interface RoomState {
  roomId: string;
  status: RoomStatus;
  players: Map<string, PlayerInfo>;
  maxPlayers: number;
  minPlayers: number;
}

// ============================================
// 메시지 타입
// ============================================

type RoomMessage =
  | { type: 'JOIN_ROOM'; playerId: string; playerName: string }
  | { type: 'LEAVE_ROOM'; playerId: string }
  | { type: 'SET_READY'; playerId: string; ready: boolean }
  | { type: 'GAME_ACTION'; playerId: string; action: any }
  | { type: 'START_GAME' }
  | { type: 'END_GAME'; winnerId?: string }
  | { type: 'GET_STATE' }
  | { type: 'TICK' };

type PlayerMessage =
  | { type: 'ROOM_STATE'; state: RoomState }
  | { type: 'PLAYER_JOINED'; player: PlayerInfo }
  | { type: 'PLAYER_LEFT'; playerId: string }
  | { type: 'PLAYER_READY'; playerId: string; ready: boolean }
  | { type: 'GAME_STARTING'; countdown: number }
  | { type: 'GAME_STARTED' }
  | { type: 'GAME_ACTION'; playerId: string; action: any }
  | { type: 'GAME_ENDED'; winnerId?: string; scores: Map<string, number> };

// ============================================
// Actor 기본 인프라
// ============================================

type ActorRef<T> = {
  send: (msg: T) => void;
  ask: <R>(msg: T) => Promise<R>;
  id: string;
};

class ActorSystem {
  private actors = new Map<string, any>();

  createActor<T, S>(
    id: string,
    initialState: S,
    behavior: (state: S, msg: T, ctx: ActorContext<T>) => S | Promise<S>
  ): ActorRef<T> {
    const actor = new StatefulActor(id, initialState, behavior, this);
    this.actors.set(id, actor);
    return actor.ref();
  }

  getActor<T>(id: string): ActorRef<T> | undefined {
    const actor = this.actors.get(id);
    return actor?.ref();
  }

  removeActor(id: string): void {
    this.actors.delete(id);
  }
}

interface ActorContext<T> {
  self: ActorRef<T>;
  system: ActorSystem;
  broadcast: (targets: ActorRef<any>[], msg: any) => void;
}

class StatefulActor<T, S> {
  private state: S;
  private mailbox: Array<{ msg: T; resolve?: (value: any) => void }> = [];
  private processing = false;
  private context: ActorContext<T>;

  constructor(
    private id: string,
    initialState: S,
    private behavior: (state: S, msg: T, ctx: ActorContext<T>) => S | Promise<S>,
    private system: ActorSystem
  ) {
    this.state = initialState;
    this.context = {
      self: this.ref(),
      system,
      broadcast: (targets, msg) => {
        targets.forEach(t => t.send(msg));
      },
    };
  }

  ref(): ActorRef<T> {
    return {
      id: this.id,
      send: (msg: T) => {
        this.mailbox.push({ msg });
        this.processNext();
      },
      ask: <R>(msg: T): Promise<R> => {
        return new Promise(resolve => {
          this.mailbox.push({ msg, resolve });
          this.processNext();
        });
      },
    };
  }

  private async processNext(): Promise<void> {
    if (this.processing) return;
    this.processing = true;

    while (this.mailbox.length > 0) {
      const { msg, resolve } = this.mailbox.shift()!;
      try {
        const result = await this.behavior(this.state, msg, this.context);
        if (result !== undefined) {
          this.state = result;
        }
        if (resolve) {
          resolve(this.state);
        }
      } catch (error) {
        console.error(`[${this.id}] Error:`, error);
      }
    }

    this.processing = false;
  }
}

// ============================================
// Room Actor
// ============================================

function createRoomActor(system: ActorSystem, roomId: string): ActorRef<RoomMessage> {
  const initialState: RoomState = {
    roomId,
    status: 'waiting',
    players: new Map(),
    maxPlayers: 4,
    minPlayers: 2,
  };

  // 플레이어 Actor 참조 저장
  const playerActors = new Map<string, ActorRef<PlayerMessage>>();

  return system.createActor<RoomMessage, RoomState>(
    `room-${roomId}`,
    initialState,
    (state, msg, ctx) => {
      switch (msg.type) {
        case 'JOIN_ROOM': {
          if (state.players.size >= state.maxPlayers) {
            console.log(`[Room ${roomId}] Room is full, rejecting ${msg.playerName}`);
            return state;
          }

          if (state.status !== 'waiting') {
            console.log(`[Room ${roomId}] Game in progress, rejecting ${msg.playerName}`);
            return state;
          }

          const player: PlayerInfo = {
            playerId: msg.playerId,
            name: msg.playerName,
            isReady: false,
            score: 0,
          };

          state.players.set(msg.playerId, player);

          // 플레이어 Actor 생성
          const playerActor = createPlayerActor(system, msg.playerId, msg.playerName);
          playerActors.set(msg.playerId, playerActor);

          console.log(`[Room ${roomId}] ${msg.playerName} joined (${state.players.size}/${state.maxPlayers})`);

          // 다른 플레이어들에게 알림
          broadcast(playerActors, msg.playerId, { type: 'PLAYER_JOINED', player });

          // 새 플레이어에게 현재 상태 전송
          playerActor.send({ type: 'ROOM_STATE', state });

          return state;
        }

        case 'LEAVE_ROOM': {
          const player = state.players.get(msg.playerId);
          if (!player) return state;

          state.players.delete(msg.playerId);
          playerActors.delete(msg.playerId);

          console.log(`[Room ${roomId}] ${player.name} left (${state.players.size}/${state.maxPlayers})`);

          // 다른 플레이어들에게 알림
          broadcast(playerActors, null, { type: 'PLAYER_LEFT', playerId: msg.playerId });

          // 게임 중이었다면 종료
          if (state.status === 'playing' && state.players.size < state.minPlayers) {
            state.status = 'finished';
            console.log(`[Room ${roomId}] Not enough players, game ended`);
            broadcast(playerActors, null, {
              type: 'GAME_ENDED',
              scores: new Map(Array.from(state.players.values()).map(p => [p.playerId, p.score])),
            });
          }

          // 룸이 비었으면 상태 리셋
          if (state.players.size === 0) {
            state.status = 'waiting';
            console.log(`[Room ${roomId}] Room is empty, resetting`);
          }

          return state;
        }

        case 'SET_READY': {
          const player = state.players.get(msg.playerId);
          if (!player || state.status !== 'waiting') return state;

          player.isReady = msg.ready;
          console.log(`[Room ${roomId}] ${player.name} is ${msg.ready ? 'ready' : 'not ready'}`);

          // 모든 플레이어에게 알림
          broadcast(playerActors, null, {
            type: 'PLAYER_READY',
            playerId: msg.playerId,
            ready: msg.ready,
          });

          // 모두 준비되었는지 확인
          checkStartCondition(state, ctx.self, playerActors);

          return state;
        }

        case 'START_GAME': {
          if (state.status !== 'waiting') return state;

          state.status = 'starting';
          console.log(`[Room ${roomId}] Game starting...`);

          // 카운트다운
          let countdown = 3;
          const countdownInterval = setInterval(() => {
            if (countdown > 0) {
              broadcast(playerActors, null, { type: 'GAME_STARTING', countdown });
              console.log(`[Room ${roomId}] Starting in ${countdown}...`);
              countdown--;
            } else {
              clearInterval(countdownInterval);
              state.status = 'playing';
              console.log(`[Room ${roomId}] Game started!`);
              broadcast(playerActors, null, { type: 'GAME_STARTED' });
            }
          }, 1000);

          return state;
        }

        case 'GAME_ACTION': {
          if (state.status !== 'playing') return state;

          const player = state.players.get(msg.playerId);
          if (!player) return state;

          console.log(`[Room ${roomId}] ${player.name} action: ${JSON.stringify(msg.action)}`);

          // 간단한 점수 시스템
          if (msg.action.type === 'SCORE') {
            player.score += msg.action.points;
            console.log(`[Room ${roomId}] ${player.name} score: ${player.score}`);
          }

          // 모든 플레이어에게 브로드캐스트
          broadcast(playerActors, null, {
            type: 'GAME_ACTION',
            playerId: msg.playerId,
            action: msg.action,
          });

          return state;
        }

        case 'END_GAME': {
          if (state.status !== 'playing') return state;

          state.status = 'finished';
          console.log(`[Room ${roomId}] Game ended! Winner: ${msg.winnerId || 'none'}`);

          const scores = new Map(
            Array.from(state.players.values()).map(p => [p.playerId, p.score])
          );

          broadcast(playerActors, null, {
            type: 'GAME_ENDED',
            winnerId: msg.winnerId,
            scores,
          });

          // 5초 후 대기 상태로 리셋
          setTimeout(() => {
            state.status = 'waiting';
            state.players.forEach(p => {
              p.isReady = false;
              p.score = 0;
            });
            console.log(`[Room ${roomId}] Room reset to waiting`);
          }, 5000);

          return state;
        }

        case 'GET_STATE': {
          return state;
        }

        default:
          return state;
      }
    }
  );
}

function broadcast(
  playerActors: Map<string, ActorRef<PlayerMessage>>,
  excludeId: string | null,
  msg: PlayerMessage
): void {
  playerActors.forEach((actor, id) => {
    if (id !== excludeId) {
      actor.send(msg);
    }
  });
}

function checkStartCondition(
  state: RoomState,
  roomActor: ActorRef<RoomMessage>,
  playerActors: Map<string, ActorRef<PlayerMessage>>
): void {
  if (state.players.size >= state.minPlayers) {
    const allReady = Array.from(state.players.values()).every(p => p.isReady);
    if (allReady) {
      console.log(`[Room ${state.roomId}] All players ready!`);
      roomActor.send({ type: 'START_GAME' });
    }
  }
}

// ============================================
// Player Actor
// ============================================

interface PlayerState {
  playerId: string;
  name: string;
  lastMessage?: PlayerMessage;
}

function createPlayerActor(
  system: ActorSystem,
  playerId: string,
  name: string
): ActorRef<PlayerMessage> {
  const initialState: PlayerState = {
    playerId,
    name,
  };

  return system.createActor<PlayerMessage, PlayerState>(
    `player-${playerId}`,
    initialState,
    (state, msg, ctx) => {
      state.lastMessage = msg;

      switch (msg.type) {
        case 'ROOM_STATE':
          console.log(`[Player ${name}] Received room state: ${msg.state.players.size} players`);
          break;

        case 'PLAYER_JOINED':
          console.log(`[Player ${name}] ${msg.player.name} joined the room`);
          break;

        case 'PLAYER_LEFT':
          console.log(`[Player ${name}] Player ${msg.playerId} left`);
          break;

        case 'PLAYER_READY':
          console.log(`[Player ${name}] Player ${msg.playerId} is ${msg.ready ? 'ready' : 'not ready'}`);
          break;

        case 'GAME_STARTING':
          console.log(`[Player ${name}] Game starting in ${msg.countdown}...`);
          break;

        case 'GAME_STARTED':
          console.log(`[Player ${name}] Game started!`);
          break;

        case 'GAME_ACTION':
          console.log(`[Player ${name}] Received action from ${msg.playerId}`);
          break;

        case 'GAME_ENDED':
          console.log(`[Player ${name}] Game ended! Winner: ${msg.winnerId || 'none'}`);
          break;
      }

      return state;
    }
  );
}

// ============================================
// 실행 예제
// ============================================

async function main() {
  console.log('=== Game Room Actor Pattern Demo ===\n');

  const system = new ActorSystem();

  // 게임 룸 생성
  const room = createRoomActor(system, '001');

  console.log('--- 1. Players Joining ---\n');

  room.send({ type: 'JOIN_ROOM', playerId: 'p1', playerName: 'Alice' });
  await delay(100);

  room.send({ type: 'JOIN_ROOM', playerId: 'p2', playerName: 'Bob' });
  await delay(100);

  room.send({ type: 'JOIN_ROOM', playerId: 'p3', playerName: 'Charlie' });
  await delay(100);

  console.log('\n--- 2. Players Getting Ready ---\n');

  room.send({ type: 'SET_READY', playerId: 'p1', ready: true });
  await delay(100);

  room.send({ type: 'SET_READY', playerId: 'p2', ready: true });
  await delay(100);

  // 마지막 플레이어가 준비되면 게임 시작
  room.send({ type: 'SET_READY', playerId: 'p3', ready: true });

  // 게임 시작 대기
  await delay(5000);

  console.log('\n--- 3. Game Actions ---\n');

  room.send({ type: 'GAME_ACTION', playerId: 'p1', action: { type: 'SCORE', points: 10 } });
  await delay(100);

  room.send({ type: 'GAME_ACTION', playerId: 'p2', action: { type: 'SCORE', points: 15 } });
  await delay(100);

  room.send({ type: 'GAME_ACTION', playerId: 'p1', action: { type: 'SCORE', points: 20 } });
  await delay(100);

  console.log('\n--- 4. Game End ---\n');

  room.send({ type: 'END_GAME', winnerId: 'p1' });
  await delay(1000);

  console.log('\n--- 5. Player Leaving ---\n');

  room.send({ type: 'LEAVE_ROOM', playerId: 'p3' });
  await delay(100);

  console.log('\n=== Demo Complete ===');
}

function delay(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

main().catch(console.error);

File: examples/04-game-room/csharp/GameRoom.cs
// Game Room Actor Pattern - 멀티플레이어 게임 룸 예제 in C#
// Orleans 스타일로 구현

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GameRoomExample
{
    // ============================================
    // 타입 정의
    // ============================================

    public enum RoomStatus
    {
        Waiting,
        Starting,
        Playing,
        Finished
    }

    public record PlayerInfo(string PlayerId, string Name, bool IsReady, int Score);

    public record RoomState(
        string RoomId,
        RoomStatus Status,
        Dictionary<string, PlayerInfo> Players,
        int MaxPlayers,
        int MinPlayers
    );

    // ============================================
    // 메시지 타입
    // ============================================

    public interface IRoomMessage { }
    public record JoinRoom(string PlayerId, string PlayerName) : IRoomMessage;
    public record LeaveRoom(string PlayerId) : IRoomMessage;
    public record SetReady(string PlayerId, bool Ready) : IRoomMessage;
    public record GameAction(string PlayerId, string ActionType, int Points) : IRoomMessage;
    public record StartGame() : IRoomMessage;
    public record EndGame(string? WinnerId) : IRoomMessage;
    public record GetState() : IRoomMessage;

    public interface IPlayerMessage { }
    public record PlayerJoined(PlayerInfo Player) : IPlayerMessage;
    public record PlayerLeft(string PlayerId) : IPlayerMessage;
    public record PlayerReadyChanged(string PlayerId, bool Ready) : IPlayerMessage;
    public record GameStarting(int Countdown) : IPlayerMessage;
    public record GameStarted() : IPlayerMessage;
    public record GameActionBroadcast(string PlayerId, string ActionType, int Points) : IPlayerMessage;
    public record GameEnded(string? WinnerId, Dictionary<string, int> Scores) : IPlayerMessage;

    // ============================================
    // Room Grain (Actor)
    // ============================================

    public interface IRoomGrain
    {
        Task<bool> JoinRoom(string playerId, string playerName);
        Task LeaveRoom(string playerId);
        Task SetReady(string playerId, bool ready);
        Task GameAction(string playerId, string actionType, int points);
        Task StartGame();
        Task EndGame(string? winnerId);
        Task<RoomState> GetState();
    }

    public class RoomGrain : IRoomGrain
    {
        private readonly string _roomId;
        private RoomStatus _status = RoomStatus.Waiting;
        private readonly Dictionary<string, PlayerInfo> _players = new();
        private readonly Dictionary<string, PlayerGrain> _playerActors = new();
        private readonly int _maxPlayers = 4;
        private readonly int _minPlayers = 2;

        public RoomGrain(string roomId)
        {
            _roomId = roomId;
        }

        public async Task<bool> JoinRoom(string playerId, string playerName)
        {
            if (_players.Count >= _maxPlayers)
            {
                Console.WriteLine($"[Room {_roomId}] Room is full, rejecting {playerName}");
                return false;
            }

            if (_status != RoomStatus.Waiting)
            {
                Console.WriteLine($"[Room {_roomId}] Game in progress, rejecting {playerName}");
                return false;
            }

            var player = new PlayerInfo(playerId, playerName, false, 0);
            _players[playerId] = player;

            var playerActor = new PlayerGrain(playerId, playerName);
            _playerActors[playerId] = playerActor;

            Console.WriteLine($"[Room {_roomId}] {playerName} joined ({_players.Count}/{_maxPlayers})");

            // 다른 플레이어들에게 알림
            await Broadcast(playerId, new PlayerJoined(player));

            return true;
        }

        public async Task LeaveRoom(string playerId)
        {
            if (!_players.TryGetValue(playerId, out var player))
                return;

            _players.Remove(playerId);
            _playerActors.Remove(playerId);

            Console.WriteLine($"[Room {_roomId}] {player.Name} left ({_players.Count}/{_maxPlayers})");

            await Broadcast(null, new PlayerLeft(playerId));

            if (_status == RoomStatus.Playing && _players.Count < _minPlayers)
            {
                _status = RoomStatus.Finished;
                Console.WriteLine($"[Room {_roomId}] Not enough players, game ended");
                await Broadcast(null, new GameEnded(null, GetScores()));
            }

            if (_players.Count == 0)
            {
                _status = RoomStatus.Waiting;
                Console.WriteLine($"[Room {_roomId}] Room is empty, resetting");
            }
        }

        public async Task SetReady(string playerId, bool ready)
        {
            if (!_players.TryGetValue(playerId, out var player) || _status != RoomStatus.Waiting)
                return;

            _players[playerId] = player with { IsReady = ready };
            Console.WriteLine($"[Room {_roomId}] {player.Name} is {(ready ? "ready" : "not ready")}");

            await Broadcast(null, new PlayerReadyChanged(playerId, ready));
            await CheckStartCondition();
        }

        public async Task GameAction(string playerId, string actionType, int points)
        {
            if (_status != RoomStatus.Playing)
                return;

            if (!_players.TryGetValue(playerId, out var player))
                return;

            Console.WriteLine($"[Room {_roomId}] {player.Name} action: {actionType}");

            if (actionType == "SCORE")
            {
                _players[playerId] = player with { Score = player.Score + points };
                Console.WriteLine($"[Room {_roomId}] {player.Name} score: {_players[playerId].Score}");
            }

            await Broadcast(null, new GameActionBroadcast(playerId, actionType, points));
        }

        public async Task StartGame()
        {
            if (_status != RoomStatus.Waiting)
                return;

            _status = RoomStatus.Starting;
            Console.WriteLine($"[Room {_roomId}] Game starting...");

            for (int countdown = 3; countdown > 0; countdown--)
            {
                await Broadcast(null, new GameStarting(countdown));
                Console.WriteLine($"[Room {_roomId}] Starting in {countdown}...");
                await Task.Delay(1000);
            }

            _status = RoomStatus.Playing;
            Console.WriteLine($"[Room {_roomId}] Game started!");
            await Broadcast(null, new GameStarted());
        }

        public async Task EndGame(string? winnerId)
        {
            if (_status != RoomStatus.Playing)
                return;

            _status = RoomStatus.Finished;
            Console.WriteLine($"[Room {_roomId}] Game ended! Winner: {winnerId ?? "none"}");

            await Broadcast(null, new GameEnded(winnerId, GetScores()));

            // 5초 후 리셋
            _ = Task.Run(async () =>
            {
                await Task.Delay(5000);
                _status = RoomStatus.Waiting;
                foreach (var pid in _players.Keys.ToList())
                {
                    _players[pid] = _players[pid] with { IsReady = false, Score = 0 };
                }
                Console.WriteLine($"[Room {_roomId}] Room reset to waiting");
            });
        }

        public Task<RoomState> GetState()
        {
            return Task.FromResult(new RoomState(
                _roomId,
                _status,
                new Dictionary<string, PlayerInfo>(_players),
                _maxPlayers,
                _minPlayers
            ));
        }

        private async Task Broadcast(string? excludeId, IPlayerMessage message)
        {
            foreach (var (id, actor) in _playerActors)
            {
                if (id != excludeId)
                {
                    await actor.Receive(message);
                }
            }
        }

        private async Task CheckStartCondition()
        {
            if (_players.Count >= _minPlayers && _players.Values.All(p => p.IsReady))
            {
                Console.WriteLine($"[Room {_roomId}] All players ready!");
                await StartGame();
            }
        }

        private Dictionary<string, int> GetScores()
        {
            return _players.ToDictionary(p => p.Key, p => p.Value.Score);
        }
    }

    // ============================================
    // Player Grain (Actor)
    // ============================================

    public class PlayerGrain
    {
        private readonly string _playerId;
        private readonly string _name;

        public PlayerGrain(string playerId, string name)
        {
            _playerId = playerId;
            _name = name;
        }

        public Task Receive(IPlayerMessage message)
        {
            switch (message)
            {
                case PlayerJoined m:
                    Console.WriteLine($"[Player {_name}] {m.Player.Name} joined the room");
                    break;
                case PlayerLeft m:
                    Console.WriteLine($"[Player {_name}] Player {m.PlayerId} left");
                    break;
                case PlayerReadyChanged m:
                    Console.WriteLine($"[Player {_name}] Player {m.PlayerId} is {(m.Ready ? "ready" : "not ready")}");
                    break;
                case GameStarting m:
                    Console.WriteLine($"[Player {_name}] Game starting in {m.Countdown}...");
                    break;
                case GameStarted:
                    Console.WriteLine($"[Player {_name}] Game started!");
                    break;
                case GameActionBroadcast m:
                    Console.WriteLine($"[Player {_name}] Received action from {m.PlayerId}");
                    break;
                case GameEnded m:
                    Console.WriteLine($"[Player {_name}] Game ended! Winner: {m.WinnerId ?? "none"}");
                    break;
            }
            return Task.CompletedTask;
        }
    }

    // ============================================
    // 메인 프로그램
    // ============================================

    public class Program
    {
        public static async Task Main()
        {
            Console.WriteLine("=== Game Room Actor Pattern Demo (C#) ===\n");

            var room = new RoomGrain("001");

            Console.WriteLine("--- 1. Players Joining ---\n");

            await room.JoinRoom("p1", "Alice");
            await Task.Delay(100);

            await room.JoinRoom("p2", "Bob");
            await Task.Delay(100);

            await room.JoinRoom("p3", "Charlie");
            await Task.Delay(100);

            Console.WriteLine("\n--- 2. Players Getting Ready ---\n");

            await room.SetReady("p1", true);
            await Task.Delay(100);

            await room.SetReady("p2", true);
            await Task.Delay(100);

            // 마지막 플레이어가 준비되면 게임 시작
            await room.SetReady("p3", true);

            // 게임 시작 대기
            await Task.Delay(1000);

            Console.WriteLine("\n--- 3. Game Actions ---\n");

            await room.GameAction("p1", "SCORE", 10);
            await Task.Delay(100);

            await room.GameAction("p2", "SCORE", 15);
            await Task.Delay(100);

            await room.GameAction("p1", "SCORE", 20);
            await Task.Delay(100);

            Console.WriteLine("\n--- 4. Game End ---\n");

            await room.EndGame("p1");
            await Task.Delay(1000);

            Console.WriteLine("\n--- 5. Player Leaving ---\n");

            await room.LeaveRoom("p3");
            await Task.Delay(100);

            Console.WriteLine("\n=== Demo Complete ===");
        }
    }
}
